components/datetime/src/metazone.rs

```diff
@@ -0,0 +1,130 @@
+    ///
+    /// ```
+    /// use icu_locid::locale;
+    /// use icu::datetime::metazone::MetaZoneCalculator;
+    ///
+    /// let provider = icu_testdata::get_provider();
+    /// let mzc = MetaZoneCalculator::new(
+    ///     locale!("en"),
+    ///     &provider,
+    /// );
+    ///
+    /// assert!(mzc.is_ok());
+    /// ```
+    pub fn new<L, ZP>(locale: L, zone_provider: &ZP) -> Result<Self, DateTimeFormatError>
+    where
+        L: Into<Locale>,
+        ZP: ResourceProvider<MetaZonePeriodV1Marker> + ?Sized,
+    {
+        let locale = locale.into();
+        let metazone_period = zone_provider
+            .load_resource(&DataRequest {
+                options: ResourceOptions::from(&locale),
+                metadata: Default::default(),
+            })?
+            .take_payload()?;
+        Ok(Self { metazone_period })
+    }
+
+    /// Calculate metazone id from timezone id and local datetime.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu::datetime::metazone::MetaZoneCalculator;
+    /// use icu_calendar::DateTime;
+    /// use icu_datetime::provider::time_zones::{MetaZoneId, TimeZoneBcp47Id};
+    /// use icu_locid::locale;
+    /// use tinystr::tinystr;
+    ///
+    /// let provider = icu_testdata::get_provider();
+    /// let mzc = MetaZoneCalculator::new(locale!("en"), &provider);
+    ///
+    /// assert_eq!(
+    ///     mzc.as_ref().unwrap().compute_metazone_from_timezone(
+    ///         TimeZoneBcp47Id(tinystr!(8, "gugum")),
+    ///         DateTime::new_iso_datetime(1969, 1, 1, 0, 0, 0).unwrap()
+    ///     ),
+    ///     None
+    /// );
+    ///
+    /// assert_eq!(
+    ///     mzc.as_ref().unwrap().compute_metazone_from_timezone(
+    ///         TimeZoneBcp47Id(tinystr!(8, "gugum")),
+    ///         DateTime::new_iso_datetime(1970, 1, 1, 0, 0, 0).unwrap()
+    ///     ),
+    ///     Some(MetaZoneId(tinystr!(4, "guam")))
+    /// );
+    ///
+    /// assert_eq!(
+    ///     mzc.as_ref().unwrap().compute_metazone_from_timezone(
+    ///         TimeZoneBcp47Id(tinystr!(8, "gugum")),
+    ///         DateTime::new_iso_datetime(1975, 1, 1, 0, 0, 0).unwrap()
+    ///     ),
+    ///     Some(MetaZoneId(tinystr!(4, "guam")))
+    /// );
+    ///
+    /// assert_eq!(
+    ///     mzc.as_ref().unwrap().compute_metazone_from_timezone(
+    ///         TimeZoneBcp47Id(tinystr!(8, "gugum")),
+    ///         DateTime::new_iso_datetime(2000, 12, 22, 15, 0, 0).unwrap()
+    ///     ),
+    ///     Some(MetaZoneId(tinystr!(4, "cham")))
+    /// );
+    /// ```
+    pub fn compute_metazone_from_timezone(
+        &self,
+        time_zone_id: TimeZoneBcp47Id,
+        local_datetime: DateTime<Iso>,
+    ) -> Option<MetaZoneId> {
+        extern crate std;
+        std::println!("contains_key0 entry");
+        if self.metazone_period.get().0.contains_key0(&time_zone_id) {
+            std::println!("contains_key0");
+            match self.metazone_period.get().0.get0(&time_zone_id) {
```

Comment:
Suggestion: You don't need to call `contains_key0` because you are calling `get0` right below that; if `get0` returns `None` then that is the same as `contains_key0` returning `false`

Reply:
sg!